book/src/cross-program-invocation.md

```diff
@@ -0,0 +1,111 @@
+    token::pay(&keyed_accounts[1..])?;
+
+    launch_missiles(keyed_accounts)?;
+}
+```
+
+The above code would require that the `token` crate be dynamically
+linked, so that a custom linker could could intercept calls and
+validate accesses to `keyed_accounts`. That is, even though the client
+intends to modify both `token` and `acme` accounts, only `token` program is
+permitted to modify the `token` account, and only the `acme` program is
+permitted to modify the `acme` account.
+
+Backing off from that ideal cross-program call, a slightly more
+verbose solution is to expose token's existing `process_instruction()`
+entrypoint to the acme program:
+
+```rust,ignore
+use token_instruction;
+
+fn launch_missiles(keyed_accounts: &[KeyedAccount]) -> Result<()> {
+    ...
+}
+
+fn pay_and_launch_missiles(keyed_accounts: &[KeyedAccount]) -> Result<()> {
+    let alice_pubkey = keyed_accounts[1].key;
+    let instruction = token_instruction::pay(&alice_pubkey);
+    process_instruction(&instruction)?;
+
+    launch_missiles(keyed_accounts)?;
+}
+```
+
+where `process_instruction()` is built into Solana's runtime and responsible
+for routing the given instruction to the `token` program via the instruction's
+`program_id` field. Before invoking `pay()`, the runtime must also ensure that
+`acme` didn't modify any accounts owned by `token`. It does this by calling
+`runtime::verify_instruction()` and then afterward updating all the `pre_*`
+variables to tentatively commit `acme`'s account modifications. After `pay()`
+completes, the runtime must again ensure that `token` didn't modify any
+accounts owned by `acme`. It should call `verify_instruction()` again, but this
+time with the `token` program ID. Lastly, after `pay_and_launch_missiles()`
+completes, the runtime must call `verify_instruction()` one more time, where it
+normally would, but using all updated `pre_*` variables.  If executing
+`pay_and_launch_missiles()` up to `pay()` made no invalid account changes,
+`pay()` made no invalid changes, and executing from `pay()` until
+`pay_and_launch_missiles()` returns made no invalid changes, then the runtime
+can transitively assume `pay_and_launch_missiles()` as whole made no invalid
+account changes, and therefore commit all account modifications.
+
+### Setting `KeyedAccount.is_signer`
+
+When `process_instruction()` is invoked, the runtime must create a new
+`KeyedAccounts` parameter using the signatures from the *original* transaction
+data. Since the `token` program is immutable and existed on-chain prior to the
+`acme` program, the runtime can safely treat the transaction signature as a
+signature of a transaction with a `token` instruction. When the runtime sees
+the given instruction references `alice_pubkey`, it looks up the key in the
+transaction to see if that key corresponds to a transaction signature. In this
+case it does and so sets `KeyedAccount.is_signer`, thereby authorizing the
+`token` program to modify Alice's account.
```

Comment:
want to add a blurb on safety?  I don't think we need to fix it yet, but basically

* Alice observes chain state A
* Bob observes chain state A
* Alice signs transaction to transition A -> B
* Bob signs a transaction to transition A -> C
* Both may end up executing in any order

Reply:
Sure, great point.  So the solution I was thinking of is that that we can simply look at the accounts modified since the transaction blockhash. If any debitable accounts have changed, we fail the transaction.  The downside, of course, is that then the only way to do a fast sequence of transactions is to batch all the instructions into a single transaction.  If that transaction is larger than max_transaction_size (like loading an on-chain program into an account), then you're stuck waiting for the next blockhash.  Perhaps this is finally where it'd make sense to add sequence numbers to transactions. tag: @mvines

Reply:
@garious Combining the instruction into a single TX won't work in this case, since Alice and Bob are not aware of each other.  

Encode the explicit instructions the program will execute. The state changes can be composed out of order, but consistency is preserved since the transaction is explicit about the state transitions it expects.

Reply:
@aeyakovenko, it works because the loser of the Alice/Bob race is forced to replay their transaction after acknowledging the chain is no longer at state A.

Encoding the explicit instructions isn't better.  That'd result in the loser explicitly telling the system to do something passed on state A rather than implicitly doing it. The right fix to publish assumptions in the transaction (the blockhash tells the system the last observed state), and then have the Bank reject transactions when the stated assumptions are no longer valid.

Reply:
> Encoding the explicit instructions isn't better. That'd result in the loser explicitly telling the system to do something based on state A rather than implicitly doing it. 

It allows more instructions per second to execute.  Checking what state the network, like what blockhash the state is at is 400ms.  

Both Bob and Alice tell the system to do A -> B, and A -> C, and the program decides if its legal to run those in any order since it has a global view over all of its dependencies.

The problem with allowing a metaprogram to do that is because

* ProgramA calls ProgramB and ProgramC

* But ProgramB and ProgramC can be independently called by users.  So ProgramA can't control all the state transitions of its dependencies.

Reply:
@jackcmay ^^^ can you work to understand @aeyakovenko's concerns here and what we can do about it?  I'd prefer not to add all potential cross-program instructions to the top-level transaction, as that'd impact programmer usability.